### PR TITLE
gddrescue: add ddrescue a data recovery tool

### PR DESCRIPTION
**Pull request by lucize**
Signed-off-by: Lucian Cristian <lucian.cristian@gmail.com>

Maintainer: me
Compile/Run tested: sunxi r6628-c089671339

### CODE REVIEW

**Comment by diizzyy** [comment]
While I'm not sure if this is this even needed as most gnu dd bugs are fixed to my knowledge can't you just use the source code repo instead?

**Comment by lucize** [comment]
@diizzyy hmm I could ser about using debian's gddrescue
I don't know if dd can resume recovery, scrapping the bad blocks, reverse reading and all other options that are available to this tool

**Comment by diizzyy** [comment]
http://www.garloff.de/kurt/linux/ddrescue/ might be worth looking into?
I'm not sure if most of these claims are still valid as conv=noerror,sync pretty much works fine these days or at least it does for me (tm)?

**Comment by lucize** [comment]
I've added the Debian source
didn't use dd_rescue and I had success using ddrescue, so having this on a small system to run recovery from is very useful, usually it takes days or weeks for the data to be recovered

**Comment by diizzyy** [comment]
Hmm.... I wonder if this is more usable?
https://sortix.org/rw/

**Comment by lucize** [comment]
maybe but ddrescue has way more options and the size is ok (113.7K and 61.6K uncompressed on sunxi)
https://www.gnu.org/software/ddrescue/manual/ddrescue_manual.html

File: utils/gddrescue/Makefile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/gddrescue
+  SECTION:=utils
+  CATEGORY:=Utilities
+  TITLE:=Data recovery tool
+  URL:=https://www.gnu.org/software/ddrescue/
+  DEPENDS:= +libstdcpp
```
Comment (neheb) [inline utils/gddrescue/Makefile]: is uclibc++ possible here?

Comment (lucize) [reply utils/gddrescue/Makefile]: I'm getting this

```
In file included from /home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/string:25:0,
                 from mapfile.cc:26:
/home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/vector: In instantiation of 'void std::vector<T, Allocator>::_insert_from_iterator(std::vector<T, Allocator>::iterator, InputIterator, InputIterator) [with InputIterator = Sblock*; T = Sblock; Allocator = std::allocator<Sblock>; std::vector<T, Allocator>::iterator = Sblock*]':
/home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/vector:288:26:   required from 'void std::vector<T, Allocator>::_dispatch_insert(std::vector<T, Allocator>::iterator, InputIterator, InputIterator, std::__false_type) [with InputIterator = Sblock*; T = Sblock; Allocator = std::allocator<Sblock>; std::vector<T, Allocator>::iterator = Sblock*]'
/home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/vector:297:20:   required from 'void std::vector<T, Allocator>::insert(std::vector<T, Allocator>::iterator, InputIterator, InputIterator) [with InputIterator = Sblock*; T = Sblock; Allocator = std::allocator<Sblock>; std::vector<T, Allocator>::iterator = Sblock*]'
mapfile.cc:331:77:   required from here
/home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/vector:270:6: error: no matching function for call to 'Sblock::Sblock()'
    T temp;
      ^~~~
In file included from mapfile.cc:30:0:
block.h:107:3: note: candidate: Sblock::Sblock(long long int, long long int, Sblock::Status)
   Sblock( const long long p, const long long s, const Status st )
   ^~~~~~
block.h:107:3: note:   candidate expects 3 arguments, 0 provided
block.h:105:3: note: candidate: Sblock::Sblock(const Block&, Sblock::Status)
   Sblock( const Block & b, const Status st )
   ^~~~~~
block.h:105:3: note:   candidate expects 2 arguments, 0 provided
block.h:95:7: note: candidate: Sblock::Sblock(const Sblock&)
 class Sblock : public Block
       ^~~~~~
block.h:95:7: note:   candidate expects 1 argument, 0 provided
Makefile:64: recipe for target 'mapfile.o' failed
make[4]: *** [mapfile.o] Error 1
```

File: utils/gddrescue/Makefile [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#
+# Copyright (C) 2018 Lucian Cristian
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gddrescue
+PKG_VERSION:=1.22
```
Comment (neheb) [inline utils/gddrescue/Makefile]: This version is no longer on Debian's mirror. Please update.

Comment (lucize) [reply utils/gddrescue/Makefile]: yes, I've tested with 1.23

File: utils/gddrescue/Makefile [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  DEPENDS:= +libstdcpp
+endef
+
+define Package/gddrescue/description
+	GNU ddrescue is a data recovery tool. It copies data from one file
+	or block device (hard disc, cdrom, etc) to another, trying to rescue
+	the good parts first in case of read errors.
+endef
+
+MAKE_FLAGS += \
```
Comment (neheb) [inline utils/gddrescue/Makefile]: More needs to be done here. From brief compilation I'm getting:

```
g++-uc  -Wall -W -O2 -c -o arg_parser.o arg_parser.cc
```

none of the CFLAGS are getting passed.

File: utils/gddrescue/Makefile (hunk 1 above)
Comment (neheb) [reply utils/gddrescue/Makefile]: The error here is due to uClibc++ not supporting long long. Sigh.

File: utils/gddrescue/Makefile (hunk 3 above)
Comment (lucize) [reply utils/gddrescue/Makefile]: added something like this

```
define Build/Compile
        $(MAKE) -C $(PKG_BUILD_DIR) \
                CC="$(TARGET_CXX)" \
                CFLAGS="$(TARGET_CFLAGS) $(TARGET_CPPFLAGS) -fno-rtti" \
                all
endef
```
but then still
```
Package gddrescue is missing dependencies for the following libraries:
libc.so.6
libstdc++.so.6

```

Comment (neheb) [reply utils/gddrescue/Makefile]: I'm getting this when trying to compile with uclibc++:

```
creating config.status
creating Makefile
VPATH = .
prefix = /usr
exec_prefix = /usr
bindir = /usr/bin
datarootdir = $(prefix)/share
infodir = /usr/info
mandir = /usr/man
CXX = g++
CPPFLAGS = 
CXXFLAGS = -Wall -W -O2
LDFLAGS = 
OK. Now you can run make.
```

CXX is then later changed to gcc+-uc by the build system but not CXX or CPPFLAGS. Or LDFLAGS.

Also, it turns out there's a configure script you can modify.

Comment (lucize) [reply utils/gddrescue/Makefile]: the values are hardcoded in configure, adding
```
CXXFLAGS = -Wall -W -O2 -flto -fno-rtti
LDFLAGS = -Wl,--as-needed -flto
```
still results in
```
Package gddrescue is missing dependencies for the following libraries:
libc.so.6
libstdc++.so.6

```

Comment (neheb) [reply utils/gddrescue/Makefile]: Here's the patch that I use:

https://gist.github.com/neheb/eacb3dbec32357f2be9cec3b1ceeda9c

It works properly. That is, it fails compilation. Until the uclibc++ issue can be figured out, we would need to stick with libstdcpp.

Comment (neheb) [reply utils/gddrescue/Makefile]: The configure script also has an interesting --enable-non-posix option which could be useful.

Comment (neheb) [reply utils/gddrescue/Makefile]: I asked upstream to see what they think: http://lists.gnu.org/archive/html/bug-ddrescue/2019-03/msg00000.html

Comment (lucize) [reply utils/gddrescue/Makefile]: ahh, it used g++
setting 

```
CXX = g++-uc
CXXFLAGS = -Wall -W -Os -flto -fno-rtti
LDFLAGS = -flto -nodefaultlibs
```

still errors with `/home/md10/build/sunxi/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/usr/include/uClibc++/vector:270:6: error: no matching function for call to 'Sblock::Sblock()'`

**Comment by lucize** [comment]
i've left the 
`#include $(INCLUDE_DIR)/uclibc++.mk`
for future if there will be a fix

**Comment by neheb** [comment]
Only annoyance is using Debian's tarball instead of the original because of missing lzip support and because it seems it doesn't get updated as often. (latest upstream is 1.24).

But no real issues. ACK from me.

**Comment by lucize** [comment]
the lzip patch is here, but was marked as bit rotten
I managed to rescue a drive with 1TB of data, it only took like one year and some months :) , so is not important to be on the latest.
maybe we can patch the diff between 1.23 and 1.24 ?

**Comment by neheb** [comment]
Well, there's this: https://github.com/openwrt/openwrt/pull/867 but supporting it properly would probably need a package in tools/.

As far as the version goes, up to you.

File: utils/gddrescue/Makefile (hunk 1 above)
Comment (neheb) [reply utils/gddrescue/Makefile]: Upstream provided a patch which will be part of 1.25 that fixes compilation with uclibc++. It's really quite simple.

https://gist.github.com/neheb/5e340598e2e9c31939a3570d106816aa

**Comment by lucize** [comment]
that fixed it, thanks !
ready to be merged

**Comment by diizzyy** [comment]
1.24 is out, might want to pull in Michael Niedermayer's commits too while at it?
https://github.com/michaelni/ddrescue/commits/master

**Comment by lucize** [comment]
ok, I'll add a patch or maybe is better to add the lzip to core tools and switch to the original source ?

**Comment by diizzyy** [comment]
A bit old but https://blogs.gentoo.org/mgorny/2014/02/22/a-few-words-on-lzip-compressor/ ...
I'm either fine with Debians tarball or if you're fine using Michael's repo instead.
I don't think we should give in on the lzip stuff to be honest.

If anything you could try to get libarchive in tools which handles lzip, zip, tar, xz  so it could replace about all current tools etc but last time I suggested libarchive it was more or less ignored.
https://github.com/libarchive/libarchive/wiki

**Comment by neheb** [comment]
Interesting...Antonio Diaz is the developer of gddrescue. Makes sense...

**Comment by neheb** [comment]
I just tested this on my Turris Omnia. Seems to work. Merging.

**Pull request merged**